Restructure user guide into top-level guides, reference and explanation parts

malteneuss opened the pull request:
Another step of the user guide improvement initiative #9214:

Restructure Cabal documentation top-level parts 

The goal is for users to easier find pages for typical problems through search engines and page navigation.
- The top-level layout is based on the popular documentation structure by https://documentation.divio.com/ to give a
   clear structure to users and future documentation contributors:
  * Guides: Present a solution to a single, atomic, typical user problem. 
  * Reference: Describe user API (CLI fields, syntax etc) with technical rigour and completeness. 
  * Explanation: Discuss background information, scope, design decisions etc.
- Move existing documentation roughly into these categories with minimal editing as the basis for further editing.
- Rename guide titles to mention how-to for improving SEO.
- Rename some files to improve SEO since that name becomes part of the URL (often called slug). 
  Important page keywords should appear in the slug as well to make pages rank higher in search engines.

**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

ffaf1 commented on doc/_templates/layout.html:
```diff
@@ -2,7 +2,7 @@
   {{ super() }}
-  <a href="cabal-projectindex.html">Reference</a>
-  <a href="genindex.html">Index</a>
+  <a href="cabal-projectindex.html">Quicklinks</a>
+  <a href="genindex.html">Word index</a>
```
I associate “Word index” to Forth. “Index” is

```
A table for facilitating reference to topics, names, and
the like, in a book, usually giving the page on which a
particular word or topic may be found; -- usually
alphabetical in arrangement, and printed at the end of the
volume. Typically found only in non-fiction books.
[1913 Webster +PJC]
```

what does “word” add to it?

malteneuss replied:
Indeed, reverted.

ffaf1 commented on doc/_templates/layout.html:
```diff
@@ -2,7 +2,7 @@
   {{ super() }}
-  <a href="cabal-projectindex.html">Reference</a>
-  <a href="genindex.html">Index</a>
+  <a href="cabal-projectindex.html">Quicklinks</a>
+  <a href="genindex.html">Word index</a>
```
On “Quicklinks” vs. “Reference” I am ambivalent, but we need to understand what exactly the section is about. It is not a TOC, nor an index for sure.

ffaf1 replied:
Ah, I see, we *already* have a `Cabal feference` section in the TOC, and this indeed seems a condensed version of `3. Cabal reference`. Let's see if people come up with a better name (“Cabal syntax quicklinks”?)

malteneuss replied:
I like "Cabal Syntax Quicklinks". Let's try this.

ffaf1 commented on doc/overview-cabal-reference.rst:
```diff
@@ -0,0 +1,15 @@
+###############
+Cabal Reference
```
Is it Quicklinks or Reference?

ffaf1 commented on doc/cabal-package-description-file.rst:
```diff
@@ -1,6 +1,8 @@
-Package Description
-===================
+package.cabal File
```
Does rst allow “`.cabal` File”?

I say for newcomers not to slip on the fact that `cabal.project` is a file with `.project` extension and `cabal` name, while `package.cabal` actually can have many names.

geekosaur replied:
Possibly `*.cabal file` here.

malteneuss replied:
How about a placeholder syntax like `<package>.cabal` to give a name to what will be there?

ffaf1 replied:
Fine by me!

ulysses4ever replied:
I dislike titles starting from file names, commands and other technicalities like that, because I think they don't help to understand what the section is about — the main point of a title. At the same time, I agree that this particular title better make it clear that the section has to do with the cabal file. So, my preference would be sometime like:

```
Package description — <package>.cabal file
```
(or whatever the proper punctuation mark is in a view of a native speaker). 

I also (like David) prefer to explicitly placeholder syntax of some form. It's be nice, in the future, to make a pass over the whole manual and harmonize our approach in this regard.

malteneuss replied:
Sounds good. What about "Project description - cabal.project file" or "Multi-package description - cabal.project file" or "Multi-package project description - cabal.project file" to be consistent for the other file title?

ulysses4ever replied:
"Project description — cabal.project file" makes most sense to me. (Be sure to use m-dashes :pray:)

malteneuss replied:
Done

ffaf1 commented on doc/conf.py:
```diff
@@ -25,7 +25,7 @@
 source_suffix = '.rst'
 source_encoding = 'utf-8-sig'
-master_doc = 'index'
+root_doc = 'index'
 
```
This is ok, but maybe add a lil description in your commit (changed in version 4.0 of Sphynx)

malteneuss replied:
I'll fix it in final commit rewrite. (I think otherwise it messes up what changes you have already looked up).

malteneuss replied:
Ok, i had to do a rebase anyway and edited the commit message. (However, i'll do a final clean up later nonetheless).

ffaf1 commented on doc/getting-started.rst:
```diff
@@ -228,4 +228,4 @@ What Next?
 some of the resources on the Haskell website's `documentation page
 <https://www.haskell.org/documentation/>`__ or read more about packages and
-Cabal on the :doc:`introduction <intro>` page.
+Cabal on the :doc:`What Cabal does <cabal-context>` page.
```
I wonder if `cabal` is more appropriate than `Cabal`.

ffaf1 replied:
OK I wondered wrong.

ffaf1 commented on doc/how-to-report-bugs.rst:
```diff
@@ -0,0 +1,9 @@
+How to report Cabal bugs and deficiencies
```
“bugs and feature requests” seems less reduntant

malteneuss replied:
Good point, fixed.

ffaf1 commented on doc/how-to-report-bugs.rst:
```diff
@@ -0,0 +1,9 @@
+Please report any flaws or feature requests in the `bug
+tracker <https://github.com/haskell/cabal/issues>`__.
+
+For general discussion or queries email the libraries mailing list
+libraries@haskell.org. There is also a development mailing list
+cabal-devel@haskell.org.
```
I understand this patch is not meant to modify content, but we need to integrate Matrix into the docs (and possibly axe `libraries`) sooner or later.

malteneuss replied:
I just saw that this information will as part of https://github.com/haskell/cabal/pull/9354/files

ffaf1 commented on doc/package-concepts.rst:
```diff
@@ -1,308 +1,3 @@
-Quickstart
-==========
-
-.. TIP::
-    If this is your first time using `cabal` you should check out the :doc:`Getting Started guide <getting-started>`.
-
```
Then it is not a quickstart, if I need a `Getting Started guide`!

ffaf1 reviewed (commented):
I appreciate the new TOC

geekosaur commented on doc/how-to-integrate-into-nix.rst:
```diff
@@ -1,5 +1,5 @@
-Nix Integration
-===============
+How to integrate Cabal into Nix
```
Do we need this? IIRC there's a ticket noting that the functionality is broken and suggesting to delete it.

malteneuss replied:
I kept it for now to show how a consistent guide naming could look like.

BinderDavid replied:
There is already an open PR for the deletion of this section (which also removes the nix integration from cabal itself) here: https://github.com/haskell/cabal/pull/9239

ulysses4ever replied:
Yes, I think, it's better to remove it ASAP (at least the docs): it's a wart that is actively harming ux.

malteneuss replied:
Done

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+.. TIP::
+    If this is your first time using `cabal` you should check out the :doc:`Getting Started guide <getting-started>`.
+
+Starting from scratch, we're going to walk you through creating a simple
```
This sounds like it _is_ the Getting Started guide.

geekosaur replied:
On second thought, given @ffaf1's comment on a deleted line, this whole thing is just relocated and this confusion was already there. We should probably still figure out what we want.

malteneuss replied:
Yeah, the idea is that the getting started is a heavily condensed version of the more detailed package guide.

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+and/or an executable. Libraries are collections of Haskell modules that
+can be re-used by other Haskell libraries and programs, while executables
+are standalone programs. Test suites can both depend on a library or be
+standalonely generated.
```
```suggestion
standalone.
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+are standalone programs. Test suites can both depend on a library or be
+standalonely generated.
+
+For the moment these are the only choices. For more complex packages
```
```suggestion
For the moment, these are the only choices. For more complex packages
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+standalonely generated.
+
+For the moment these are the only choices. For more complex packages
+(e.g. a library and multiple executables) the ``.cabal``
```
```suggestion
(e.g. a library and multiple executables), the ``.cabal``
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+file can be edited afterwards.
+
+After you make your selection (executable; library; library
+and executable; or: test suite) cabal asks us a number of questions starting with
```
```suggestion
and executable; or: test suite), cabal asks you a number of questions starting with
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+
+After you make your selection (executable; library; library
+and executable; or: test suite) cabal asks us a number of questions starting with
+which version of the cabal specification to use, our package's name
```
```suggestion
which version of the cabal specification to use, your package's name
```
Alternately change "after you make your selection" above; these should match instead of suddenly switching from second to first person.

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+Running the program
+-------------------
+
+Now that we have our Haskell code and the extra files that Cabal needs we
```
```suggestion
Now that we have our Haskell code and the extra files that Cabal needs, we
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+   ...
+   Hello, Haskell!
+
+Since we have an executable we can use ``cabal run proglet`` which will build
```
```suggestion
Since we have an executable, we can use ``cabal run proglet`` which will build
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+.. highlight:: cabal
+
+Load up the ``.cabal`` file in a text editor. The first part of the
+``.cabal`` file has the package metadata and towards the end of the file
```
```suggestion
``.cabal`` file has the package metadata, and towards the end of the file
```
You may alos prefer to say "followed by" instead of "towards the end of the file".

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+While your library or executable may include a number of modules, it
+almost certainly also imports a number of external modules from the
+standard libraries or other pre-packaged libraries. (These other
+libraries are of course just Cabal packages that contain a library.)
```
```suggestion
libraries are of course just Cabal packages that contain one or more libraries.)
```

geekosaur commented on doc/how-to-package-haskell-code.rst:
```diff
@@ -0,0 +1,291 @@
+-  ``pkgname >=n``
+-  ``pkgname ^>=n`` (since Cabal 2.0)
+-  ``pkgname >=n && <m``
+-  ``pkgname ==n.*`` (since Cabal 1.6)
```
Do we care about the "since" here?

malteneuss replied:
Not, sure, but i like to add more explicit examples in a later PR. Let's keep it for now as is.

ulysses4ever replied:
If it's not the reference part of the manual, things like `since Cabal 1.6` are better omitted. Of course, if you feel like leaving it as a future work, that's fine.

malteneuss replied:
Removed.

malteneuss commented:
I've cleaned up the commits

ffaf1 commented on doc/cabal-package-description-file.rst:
```diff
@@ -1,6 +1,8 @@
-Package Description
-===================
+Package Description — <package>.cabal File
```
```suggestion
Package Description — ``<package>.cabal`` File
```

ffaf1 replied:
on second thought no, it is fine as it is

ffaf1 commented on doc/cabal-project-description-file.rst:
```diff
@@ -1,5 +1,5 @@
-cabal.project Reference
-=======================
+Project Description — cabal.project File
```
```suggestion
Project Description — ``cabal.project`` File
```

ffaf1 reviewed (commented):
edit: not a good idea, ignore this

---

Can you try this for extra clarity?
If it is a mess you can `reset` the commit.

Comparison:
- ![image](https://github.com/haskell/cabal/assets/87714215/7c6e1aec-98fa-453e-a5d5-5328e4a24bef)
- (tbd)

ffaf1 reviewed (approved):
The new TOC with 4+2 main sections is very good.

ulysses4ever reviewed (commented):
I like the overall direction, but it’s a little hard to navigate all the changes. Please, expand the description of the PR (+ the commit message, which I haven’t checked, but should be identical to the PR description), and explain what exactly this achieves. References to 3rd party sites is allowed but the description should be self-contained nevertheless. Right now it’s not.

one clear flaw in the current PR is renaming of the source files. Let’s probably try to avoid this since it’ll break a lot of links to the `latest` version of the docs. The name of the file doesn’t need to match the name of the section.

ulysses4ever commented:
Since it’s a significant restructure of our main documentation source, I invite the main contributors to chime In @Kleidukos @Mikolaj @fgaz @gbaz @andreasabel @andreabedini

BinderDavid commented:
I really like the new TOC! I played around a bit with using captions, i.e. using the following `index.rst`:

```

Welcome to the Cabal User Guide
===============================

.. toctree::
   :caption: Getting Started
   :numbered:

   getting-started

.. toctree::
   :caption: Cabal Guide
   :numbered:

   how-to-package-haskell-code
   how-to-build-like-nix
   how-to-integrate-into-nix
   how-to-report-bugs

.. toctree::
   :caption: Cabal Reference
   :numbered:

   cabal-package-description-file
   cabal-project-description-file
   cabal-config-and-commands
   setup-commands
   file-format-changelog
   buildinfo-fields-reference

.. toctree::
   :caption: Cabal Explanation
   :numbered:

   cabal-context
   package-concepts
   cabal-interface-stability
```
This yields the following look:
![Screenshot 2023-10-30 at 08-14-26 Welcome to the Cabal User Guide — Cabal 3 11 0 0 User's Guide](https://github.com/haskell/cabal/assets/18737248/edc130ca-4c18-4b61-b044-e4a9358afff2)

The problem is that the sections aren't consecutively numbered on the left, but this layout might be an alternative.

Mikolaj commented:
I love the restructure. 

BTW, I wonder how the structure in the aborted project https://github.com/haskell/cabal-userguide compares with this one? Does it correspond to just the "Cabal Guilde" part? Is there anything to be learned or scavenged? @JonathanLorimer, would you have any remarks about the new restructure proposed in this ticket or generally an advice from a veteran?

ulysses4ever commented:
I like David's version more. There's no reason to number the highest-level sections like Guide / Reference / Explanation in my opinion.

malteneuss commented:
> I really like the new TOC! I played around a bit with using captions, i.e. using the following `index.rst`:
> ...
> The problem is that the sections aren't consecutively numbered on the left, but this layout might be an alternative.

That's a really nice layout. I've copied it over.

malteneuss commented:
> I like the overall direction, but it’s a little hard to navigate all the changes. Please, expand the description of the PR (+ the commit message, which I haven’t checked, but should be identical to the PR description), and explain what exactly this achieves. References to 3rd party sites is allowed but the description should be self-contained nevertheless. Right now it’s not.
> 
> one clear flaw in the current PR is renaming of the source files. Let’s probably try to avoid this since it’ll break a lot of links to the `latest` version of the docs. The name of the file doesn’t need to match the name of the section.

I've expanded the PR message and squashed the relevant commits to use that same message.

Regarding file renames i've added SEO as the reason. Our aim should be to have our latest Cabal pages rank highest for typical search queries like "how to package Haskell", "cabal file", and related variations instead of GHC, stack or blogs.

malteneuss commented:
> I love the restructure.
> 
> BTW, I wonder how the structure in the aborted project https://github.com/haskell/cabal-userguide compares with this one? Does it correspond to just the "Cabal Guilde" part? Is there anything to be learned or scavenged? @JonathanLorimer, would you have any remarks about the new restructure proposed in this ticket or generally an advice from a veteran?

The structure of that alternative looks like what the authors of https://documentation.divio.com/ would call a tutorial, teaching a user step-by-step more and more involved usages of Cabal. This may not be needed with Cabal, although some the topics would make good guides as well.

BinderDavid commented:
I said something similar in #9214 already, but I think that this is an absolutely essential step in improving the documentation of cabal. Large sections in the manual can currently not decide whether they want to be a tutorial or guide introducing how to accomplish some task, or whether they want to be the reference documentation which exhaustively enumerates all the knobs and options.  Making this explicit, like this PR does, is helpful in spotting and naming these discrepancies and fixing them in subsequent smaller PRs. Thanks @malteneuss  for doing this :)

julialongtin commented:
hmm. was just trying to use these docs (i have an upstream package with a bug in it). i uploaded it to a remote git repository, and wanted to tell cabal to use that version instead of the one from hackage, but the language around `Specifying Packages from Remote Version Control Locations` left me confused. "where do i put the name of the package?"

i think "specifying packages from an external version control" was missing an action verb. "specifying packages for cabal to download from an external version control" would have been clearer, along with some language about cabal matching the name of the package you're trying to use by first downloading the package, then looking at the cabal file there.

malteneuss commented:
> hmm. was just trying to use these docs (i have an upstream package with a bug in it). i uploaded it to a remote git repository, and wanted to tell cabal to use that version instead of the one from hackage, but the language around `Specifying Packages from Remote Version Control Locations` left me confused. "where do i put the name of the package?"
> 
> i think "specifying packages from an external version control" was missing an action verb. "specifying packages for cabal to download from an external version control" would have been clearer, along with some language about cabal matching the name of the package you're trying to use by first downloading the package, then looking at the cabal file there.

This is out of scope for this PR, but such question should definitely be answered by the docs. Could you add a separate PR against the existing docs? Maybe also adding what other "type" of version control is supported? Ideally the phrasing would use keywords that you would use to search with when you had no idea about the docs.

ulysses4ever commented:
I just noticed that this patch did break URLs like `https://cabal.readthedocs.io/en/X/cabal-project.html`. Compare:

- https://cabal.readthedocs.io/en/stable/cabal-project.html
- https://cabal.readthedocs.io/en/latest/cabal-project.html

It may be good to look into restoring them using some Sphynx magic (there should be a way to assign URLs, I believe).

malteneuss commented:
I can take a look, probably on the weekend. Thanks for reporting.

ulysses4ever commented:
@malteneuss that would be awesome, thank you!
